### PR TITLE
∞-Scroll-Table

### PR DESCRIPTION
**Pull request by dlrice**
Uses several react-virtualized components:
* MultiGrid: allows rows and columns to be frozen.
* CellMeasurer: calculates the height of each cell given the width.
* AutoSizer: calculates the width and height of the table to fill the entire containing parent.
* InfiniteLoader: requests additional rows of data upon reaching the end of the table.

Three main parts of this PR:
* `InfiniteDataTableDemo` showcases the InfiniteDataTable by listening to requests to fetch more data and synthesizes from the `lorem-ipsum` library.
* `InfiniteDataTablePreprocessor` is used to prepend the data and columns being passed into InfiniteDataTableCore based on consumer choices (ie showRowNumbers, selectable, showHeader).
* `InfiniteDataTableCore` uses the react-virtualized components and provides the cellRenderer method which given a columnIndex and rowIndex returns the contents of that cell.

In addition, the `default-page-layout` has been refactored to include the header as a CSS grid element.

### CODE REVIEW

File: src/components/infinite-data-table.jsx [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+      </div>
+    );
+  }
+
+  static getNumberCell({
+    style, className, rowIndex, showHeader,
+  }) {
+    return (
+      <div className={className} style={style}>
+        {rowIndex + 1 - showHeader}
```
Comment (xwatkins) [inline src/components/infinite-data-table.jsx]: Use of a boolean as a Number

File: src/components/infinite-data-table.jsx [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+    const {
+      columns,
+      data,
+      onLoadMoreRows,
+      showHeader,
+      fixedColumnCount,
+      fixedRowCount,
+      totalNumberRows,
+    } = this.props;
+    const rowCount = showHeader + Math.min(totalNumberRows, data.length);
```
Comment (xwatkins) [inline src/components/infinite-data-table.jsx]: use of `showHeader` as a number

File: src/components/infinite-data-table.jsx [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+  }) {
+    if (showHeader && rowIndex === 0) {
+      return InfiniteDataTableCore.getHeaderCell({
+        column,
+        onHeaderClick,
+        style,
+      });
+    }
+
+    let className = `table-data ${rowIndex % 2 ? 'table-row-odd' : 'table-row-even'}`;
```
Comment (xwatkins) [inline src/components/infinite-data-table.jsx]: BEM convention for modifiers is `--` eg. `table-row--odd`

File: src/components/infinite-data-table.jsx [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+        className,
+        style,
+      });
+    }
+
+    const id = row[idKey];
+    const checked = id in selectedRows;
+
+    if (checked) {
+      className += ' table-row-selected';
```
Comment (xwatkins) [inline src/components/infinite-data-table.jsx]: BEM

File: src/components/infinite-data-table.jsx [hunk 5]
```diff
@@ -0,0 +1,427 @@
...
+        showHeader,
+      });
+    }
+
+    if (!('render' in column)) {
+      return InfiniteDataTableCore.getNoRenderCell({ style, column });
+    }
+
+    return (
+      <div className={`${className} ${checked && ' table-row-selected'}`} style={style}>
```
Comment (xwatkins) [inline src/components/infinite-data-table.jsx]: This seems to be added twice. (L187)

File: src/components/infinite-data-table.jsx [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+  CellMeasurer,
+  CellMeasurerCache,
+  MultiGrid,
+  AutoSizer,
+  InfiniteLoader,
+} from 'react-virtualized';
+import { serializableDeepAreEqual } from '../utils';
+import '../styles/components/data-table.scss';
+
+const DEFAULT_WIDTH = 200;
```
Comment (xwatkins) [inline src/components/infinite-data-table.jsx]: Are these table width/height or cell? Please add to variable name

File: src/components/infinite-data-table.jsx [hunk 7]
```diff
@@ -0,0 +1,427 @@
...
+        threshold={1}
+      >
+        {({ onRowsRendered, registerChild }) => (
+          <AutoSizer>
+            {({ width, height }) => (
+              <div className="data-table-infinite">
+                <MultiGrid
+                  cellRenderer={this.cellRenderer}
+                  columnCount={columns.length}
+                  width={width || 100}
```
Comment (xwatkins) [inline src/components/infinite-data-table.jsx]: These default width/height could be set in constants

File: src/components/infinite-data-table.jsx [hunk 8]
```diff
@@ -0,0 +1,427 @@
...
+  }) {
+    const checked = id in selectedRows;
+    return (
+      <div className={`${className} ${checked && 'table-row-selected'}`} style={style}>
+        <input type="checkbox" onChange={() => onSelect(id)} checked={id in selectedRows} />
+      </div>
+    );
+  }
+
+  static getNumberCell({
```
Comment (xwatkins) [inline src/components/infinite-data-table.jsx]: Would a name like `getRowCountCell` be more explicit? I initially thought this was a function to render cells which had numbers in them.

File: src/components/infinite-data-table.jsx [hunk 9]
```diff
@@ -0,0 +1,427 @@
...
+    );
+  }
+
+  static getSelectCell({
+    id, onSelect, style, className, selectedRows,
+  }) {
+    const checked = id in selectedRows;
+    return (
+      <div className={`${className} ${checked && 'table-row-selected'}`} style={style}>
+        <input type="checkbox" onChange={() => onSelect(id)} checked={id in selectedRows} />
```
Comment (xwatkins) [inline src/components/infinite-data-table.jsx]: You already have the value of `checked` on L122

File: src/components/infinite-data-table.jsx [hunk 10]
```diff
@@ -0,0 +1,427 @@
...
+        for (let columnIndex = 0; columnIndex < columns.length; columnIndex += 1) {
+          this.cache.clear(rowIndex, columnIndex);
+        }
+      }
+    }
+  }
+
+  getHeader() {
+    const { columns } = this.props;
+    return columns.map(column => ({ label: column.label || '', render: x => x }));
```
Comment (xwatkins) [inline src/components/infinite-data-table.jsx]: I'm not sure what `x` is. Is there a more explicit name?

File: src/components/infinite-data-table.jsx [hunk 11]
```diff
@@ -0,0 +1,427 @@
...
+  getColumnWidth({ index }) {
+    const { columns } = this.props;
+    const column = columns[index];
+    return column && column.width ? column.width : DEFAULT_WIDTH;
+  }
+
+  static getSortableHeaderCell({ column, onHeaderClick, style }) {
+    let className = 'table-head ';
+    if (column.sorted) {
+      className += column.sorted === 'ascend' ? 'table-header-ascend' : 'table-header-descend';
```
Comment (xwatkins) [inline src/components/infinite-data-table.jsx]: BEM

File: src/components/infinite-data-table.jsx (hunk 1 above)
Comment (dlrice) [reply src/components/infinite-data-table.jsx]: Done.

File: src/components/infinite-data-table.jsx (hunk 2 above)
Comment (dlrice) [reply src/components/infinite-data-table.jsx]: Done.

File: src/components/infinite-data-table.jsx (hunk 5 above)
Comment (dlrice) [reply src/components/infinite-data-table.jsx]: Good catch. Done.

File: src/components/infinite-data-table.jsx (hunk 6 above)
Comment (dlrice) [reply src/components/infinite-data-table.jsx]: Done.

File: src/components/infinite-data-table.jsx (hunk 7 above)
Comment (dlrice) [reply src/components/infinite-data-table.jsx]: Done.

File: src/components/infinite-data-table.jsx (hunk 8 above)
Comment (dlrice) [reply src/components/infinite-data-table.jsx]: Done.

File: src/components/infinite-data-table.jsx (hunk 9 above)
Comment (dlrice) [reply src/components/infinite-data-table.jsx]: Good catch. Done.

File: src/components/infinite-data-table.jsx (hunk 10 above)
Comment (dlrice) [reply src/components/infinite-data-table.jsx]: `getHeader` is a relic from a previous life. Removed the method.

File: src/components/infinite-data-table.jsx (hunk 5 above)
Comment (dlrice) [reply src/components/infinite-data-table.jsx]: Done.

File: src/components/infinite-data-table.jsx (hunk 7 above)
Comment (dlrice) [reply src/components/infinite-data-table.jsx]: Done.

File: src/components/infinite-data-table.jsx (hunk 3 above)
Comment (dlrice) [reply src/components/infinite-data-table.jsx]: Done.

File: src/components/infinite-data-table.jsx (hunk 4 above)
Comment (dlrice) [reply src/components/infinite-data-table.jsx]: Done.

File: src/components/infinite-data-table.jsx (hunk 11 above)
Comment (dlrice) [reply src/components/infinite-data-table.jsx]: Done.

**Pull request merged**